File: solutions/Stack/84_largest-rectangle-in-histogram.py
from typing import List

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        res = 0
        stk = []

        for ind, val in enumerate(heights):
            i = ind
            while stk and stk[-1][1] > val:
                i, x = stk.pop()
                res = max(res, x * (ind - i))

            stk.append((i, val))

        while stk:
            i, x = stk.pop()
            res = max(res, x * (len(heights) - i))

        return res
    
# Time Complexity: O(n)
# Space Complexity: O(n)

File: C++/Stack/84_largest-rectangle-in-histogram.cpp
#include <stack>
#include <vector>

using namespace std;

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        size_t n = heights.size();
        stack<pair<int, int>> stk;
        int ind, i, val, res = 0;
        
        for (int i = 0; i < n; ++i) {
            ind = i;
            while (!stk.empty() && stk.top().second > heights[i]) {
                ind = stk.top().first;
                val = stk.top().second;
                stk.pop();
                res = max(res, val * (i - ind));
            }

            stk.push({ind, heights[i]});
        }

        while (!stk.empty()) {
            i = stk.top().first;
            val = stk.top().second;
            stk.pop();
            res = max(res, val * ((int)n - i));
        }

        return res;
    }
};